drivers/mspi/mspi_ambiq_ap4.c

```diff
@@ -0,0 +1,1530 @@
+			ret = -ENOTSUP;
+			goto e_return;
+		}
+
+		if ((param_mask & (~(MSPI_DEVICE_CONFIG_FREQUENCY | MSPI_DEVICE_CONFIG_IO_MODE |
+				     MSPI_DEVICE_CONFIG_CE_NUM | MSPI_DEVICE_CONFIG_DATA_RATE |
+				     MSPI_DEVICE_CONFIG_CMD_LEN | MSPI_DEVICE_CONFIG_ADDR_LEN)))) {
+			LOG_INST_ERR(cfg->log, "%u, config type not supported.", __LINE__);
+			ret = -ENOTSUP;
+			goto e_return;
+		}
+
+		if (param_mask & MSPI_DEVICE_CONFIG_FREQUENCY) {
+			hal_dev_cfg.eClockFreq = mspi_set_freq(cfg, dev_cfg->freq);
+			if (hal_dev_cfg.eClockFreq == 0) {
+				ret = -ENOTSUP;
+				goto e_return;
+			}
+			ret = am_hal_mspi_control(data->mspiHandle, AM_HAL_MSPI_REQ_CLOCK_CONFIG,
+						  &hal_dev_cfg.eClockFreq);
+			if (ret) {
+				LOG_INST_ERR(cfg->log, "%u, failed to configure eClockFreq.",
+					     __LINE__);
+				ret = -EHOSTDOWN;
+				goto e_return;
+			}
+			data->dev_cfg.freq = dev_cfg->freq;
+		}
+
+		if ((param_mask & MSPI_DEVICE_CONFIG_IO_MODE) ||
+		    (param_mask & MSPI_DEVICE_CONFIG_CE_NUM) ||
+		    (param_mask & MSPI_DEVICE_CONFIG_DATA_RATE)) {
+			enum mspi_io_mode io_mode;
+			enum mspi_data_rate data_rate;
+			uint8_t ce_num;
+
+			if (param_mask & MSPI_DEVICE_CONFIG_IO_MODE) {
+				io_mode = dev_cfg->io_mode;
+			} else {
+				io_mode = data->dev_cfg.io_mode;
+			}
+
+			if (param_mask & MSPI_DEVICE_CONFIG_DATA_RATE) {
+				data_rate = dev_cfg->data_rate;
+			} else {
+				data_rate = data->dev_cfg.data_rate;
+			}
+
+			if (param_mask & MSPI_DEVICE_CONFIG_CE_NUM) {
+				ce_num = dev_cfg->ce_num;
+			} else {
+				ce_num = data->dev_cfg.ce_num;
+			}
+
+			hal_dev_cfg.eDeviceConfig = mspi_set_line(cfg, io_mode, data_rate, ce_num);
+
+			if (hal_dev_cfg.eDeviceConfig == AM_HAL_MSPI_FLASH_MAX) {
+				ret = -ENOTSUP;
+				goto e_return;
+			}
+
+			ret = am_hal_mspi_control(data->mspiHandle, AM_HAL_MSPI_REQ_DEVICE_CONFIG,
+						  &hal_dev_cfg.eDeviceConfig);
+			if (ret) {
+				LOG_INST_ERR(cfg->log, "%u, failed to configure device.", __LINE__);
+				ret = -EHOSTDOWN;
+				goto e_return;
+			}
+
+			data->dev_cfg.io_mode = dev_cfg->io_mode;
+			data->dev_cfg.data_rate = dev_cfg->data_rate;
+			data->dev_cfg.ce_num = dev_cfg->ce_num;
+		}
+
+		if (param_mask & MSPI_DEVICE_CONFIG_CMD_LEN) {
+			if (dev_cfg->cmd_length > AM_HAL_MSPI_INSTR_2_BYTE + 1 ||
+			    dev_cfg->cmd_length == 0) {
+				LOG_INST_ERR(cfg->log, "%u, invalid cmd_length.", __LINE__);
+				ret = -ENOTSUP;
+				goto e_return;
+			}
+			hal_dev_cfg.eInstrCfg = dev_cfg->cmd_length - 1;
+			data->dev_cfg.cmd_length = dev_cfg->cmd_length;
+		}
+
+		if (param_mask & MSPI_DEVICE_CONFIG_ADDR_LEN) {
+			if (dev_cfg->addr_length > AM_HAL_MSPI_ADDR_4_BYTE + 1 ||
+			    dev_cfg->addr_length == 0) {
+				LOG_INST_ERR(cfg->log, "%u, invalid addr_length.", __LINE__);
+				ret = -ENOTSUP;
+				goto e_return;
+			}
+			hal_dev_cfg.eAddrCfg = dev_cfg->addr_length - 1;
+			data->dev_cfg.addr_length = dev_cfg->addr_length;
+		}
+	} else {
+		am_hal_mspi_dqs_t *dqsCfg = &g_sMspiDqsCfg;
```

Comment:
This type of programming is buggy. This mspi_ambiq_ap4 will be used by all mspi instances and the content of g_sMspiDqsCfg keeps getting changed every time the mspi_ambiq_dev_config gets called even if it is from a different mspi instance. 
Should place this in struct mspi_ambiq_data